README.txt
NAME

flexnbd - A fast NBD server

SYNOPSIS

  flexnbd MODE [ ARGS ]

  flexnbd serve --addr ADDR --port PORT --file FILE [--sock SOCK]
    [--default-deny] [--killswitch] [global_option]* [acl_entry]*

  flexnbd listen --addr ADDR --port PORT --file FILE [--sock SOCK]
    [--default-deny] [global_option]* [acl_entry]*

  flexnbd mirror --addr ADDR --port PORT --sock SOCK [--unlink]
    [--bind BIND_ADDR] [global_option]*

  flexnbd acl --sock SOCK [acl_entry]+ [global_option]*

  flexnbd break --sock SOCK [global_option]*

  flexnbd status --sock SOCK [global_option]*

  flexnbd read --addr ADDR --port PORT --from OFFSET --size SIZE
    [--bind BIND_ADDR] [global_option]*

  flexnbd write --addr ADDR --port PORT --from OFFSET --size SIZE
    [--bind BIND_ADDR] [global_option]*

  flexnbd help [mode] [global_option]*

DESCRIPTION

Flexnbd is a fast NBD server which supports live migration. Live
migration is performed by writing the data to a new server. A failed
migration will be invisible to any connected clients.

Flexnbd tries quite hard to preserve sparsity of files it is serving,
even across migrations.

SERVE MODE

Serve a file.

  $ flexnbd serve --addr <ADDR> --port <PORT> --file <FILE>
      [--sock <SOCK>] [--default-deny] [-k] [global_option]*
      [acl_entry]*

If any ACL entries are given (which should be IP
addresses), only those clients listed will be permitted to connect.

flexnbd will continue to serve until a SIGINT, SIGQUIT, or a successful
migration.

  OPTIONS

  --addr, -l ADDR  
    The address to listen on. Required.

  --port, -p PORT  
    The port to listen on. Required.

  --file, -f FILE  
    The file to serve. Must already exist. Required.

  --sock, -s SOCK  
    Path to a control socket to open. You will need this if you want to
    migrate, get the current status, or manipulate the access control
    list.

  --default-deny, -d  
    How to interpret an empty ACL. If --default-deny is given, an
    empty ACL will let no clients connect. If it is not given, an
    empty ACL will let any client connect.

  --killswitch, -k  
    If set, we implement a 2-minute timeout on NBD requests and
    responses. If a request takes longer than that to complete,
    the client is disconnected. This is useful to keep broken
    clients from breaking migrations, among other things.

LISTEN MODE

Listen for an inbound migration, and quit with a status of 0 on
completion.

  $ flexnbd listen --addr ADDR --port PORT --file FILE
      [--sock SOCK] [--default-deny] [global_option]* 
      [acl_entry]*

flexnbd will wait for a successful migration, and then quit. The file
to write the inbound migration data to must already exist before you
run 'flexnbd listen'.

Only one sender may connect to send data, and if the sender
disconnects part-way through the migration, the destination will
expect it to reconnect and retry the whole migration. It isn't safe
to assume that a partial migration can be resumed because the
destination has no knowledge of whether a client has made a write to
the source in the interim.

If the migration fails for a reason which the 'flexnbd listen' process
can't fix (say, a failed local write), it will exit with an error
status. In this case, the sender will continually retry the migration
until it succeeds, and you will need to restart the 'flexnbd listen'
process to allow that to happen.

  OPTIONS

As for serve.

MIRROR MODE

Start a migration from the server with control socket SOCK to the server
listening at ADDR:PORT.

  $ flexnbd mirror --addr ADDR --port PORT --sock SOCK [--unlink]
      [--bind BIND_ADDR] [global_option]*

Migration can be a slow process. Rather than block the 'flexnbd mirror'
process until it completes, it will exit with a message of "Migration
started" once it has confirmation that the local server was able to
connect to ADDR:PORT and got an NBD header back. To check on the
progress of a running migration, use 'flexnbd status'.

If the destination unexpectedly disconnects part-way through the
migration, the source will attempt to reconnect and start the migration
again. It is not safe to resume the migration from where it left off
because the source can't see that the backing store behind the
destination is intact, or even on the same machine.

If the --unlink option is given, the local file will be deleted
immediately before the mirror connection is terminated. This allows
an otherwise-ambiguous situation to be resolved: if you don't unlink
the file and the flexnbd process at either end is terminated, it's not
possible to tell which copy of the data is canonical. Since the
unlink happens as soon as the sender knows that it has transmitted all
the data, there can be no ambiguity.

Note: files smaller than 4096 bytes cannot be mirrored.

  OPTIONS

  --addr, -l ADDR  
    The address of the remote server to migrate to. Required.

  --port, -p PORT  
    The port of the remote server to migrate to. Required.

  --sock, -s SOCK  
    The control socket of the local server to migrate from. Required.

  --unlink, -u  
    Unlink the served file from the local filesystem after
    successfully mirroring.

  --bind, -b BIND_ADDR  
    The local address to bind to. You may need this if the remote
    server is using an access control list.

BREAK MODE

Stop a running migration.

  $ flexnbd break --sock SOCK [global_option]*

  OPTIONS

  --sock, -s SOCK  
    The control socket of the local server whose migration to stop.
    Required.

ACL MODE

Set the access control list of the server with the control socket SOCK
to the given access control list entries.

  $ flexnbd acl --sock SOCK [acl_entry]+ [global_option]*

ACL entries are given as IP addresses.

  OPTIONS

  --sock, -s SOCK  
    The control socket of the server whose ACL to replace. Required

STATUS MODE

Get the current status of the server with control socket SOCK.

  $ flexnbd status --sock SOCK [global_option]*

The status will be printed to STDOUT. It is a space-separated list of
key=value pairs. The space character will never appear in a key or
value. Currently reported values are:

pid  
  The process id of the server listening on SOCK.

is_mirroring  
  'true' if this server is sending migration data, 'false' otherwise.

has_control  
  'false' if this server was started in 'listen' mode. 'true' otherwise.

  OPTIONS

  --sock, -s SOCK  
    The control socket of the server of interest. Required.

READ MODE

Connect to the server at ADDR:PORT, and read SIZE bytes starting at
OFFSET in a single NBD query.

  $ flexnbd read --addr ADDR --port PORT --from OFFSET --size SIZE
      [--bind BIND_ADDR] [global_option]*

The returned data will be echoed to STDOUT. In case of a remote ACL,
set the local source address to BIND_ADDR.

  OPTIONS

  --addr, -l ADDR  
    The address of the remote server. Required.

  --port, -p PORT  
    The port of the remote server. Required.

  --from, -F OFFSET  
    The byte offset to start reading from. Required.  Maximum 2^62.

  --size, -S SIZE  
    The number of bytes to read. Required. Maximum 2^30.

  --bind, -b BIND_ADDR  
    The local address to bind to. You may need this if the remote
    server is using an access control list.

WRITE MODE

Connect to the server at ADDR:PORT, and write SIZE bytes from STDIN
starting at OFFSET in a single NBD query.

  $ cat ... | flexnbd write --addr ADDR --port PORT --from OFFSET
      --size SIZE  [--bind BIND_ADDR] [global_option]*

In case of a remote ACL, set the local source address to BIND_ADDR.

  OPTIONS

  --addr, -l ADDR  
    The address of the remote server. Required.

  --port, -p PORT  
    The port of the remote server. Required.

  --from, -F OFFSET  
    The byte offset to start writing from. Required.  Maximum 2^62.

  --size, -S SIZE  
    The number of bytes to write. Required. Maximum 2^30.

  --bind, -b BIND_ADDR  
    The local address to bind to. You may need this if the remote
    server is using an access control list.

HELP MODE

  $ flexnbd help [mode] [global_option]*

Without mode, show the list of available modes. With mode, show help for that mode.

GLOBAL OPTIONS

--help, -h  Show mode or global help.

--verbose, -v  Output all available log information to STDERR.

--quiet, -q  Output as little log information as possible to STDERR.

LOGGING

Log output is sent to STDERR. If --quiet is set, no output will be
seen unless the program termintes abnormally. If neither --quiet nor
--verbose are set, no output will be seen unless something goes wrong
with a specific request. If --verbose is given, every available log
message will be seen (which, for a debug build, is many). It is not an
error to set both --verbose and --quiet. The last one wins.

The log line format is:

  <TIMESTAMP>:<LEVEL>:<PID> <THREAD> <SOURCEFILE:SOURCELINE>: <MSG>

<TIMESTAMP>  
  Time the log entry was made. This is expressed in terms of monotonic
  ms.

<LEVEL>  
  This will be one of 'D', 'I', 'W', 'E', 'F' in increasing order of
  severity. If flexnbd is started with the --quiet flag, only 'F'
  will be seen. If it is started with the --verbose flag, any from 'I'
  upwards will be seen. Only if you have a debug build and start it
  with --verbose will you see 'D' entries.

<PID>  
  This is the process ID.

<THREAD>  
  There are several pthreads per flexnbd process: a main thread, a
  serve thread, a thread per client, and possibly a pair of mirror
  threads and a control thread. This field identifies which thread was
  responsible for the log line.

<SOURCEFILE:SOURCELINE>  
  Identifies where in the source code this log line can be found.

<MSG>  
  A short message describing what's happening, how it's being done, or
  if you're very lucky why it's going on.

EXAMPLES

  SERVING A FILE

The simplest case is serving a file on the default nbd port:

  $ cp /etc/passwd /tmp
  $ flexnbd serve --file /tmp/passwd --addr 0.0.0.0 --port 4777 &
  $ flexnbd read --addr 127.0.0.1 --port 4777 --from 0 --size 7
  root:x:
  $

  READING SERVER STATUS

In order to read a server's status, we need it to open a control socket.

  $ flexnbd serve --file /tmp/passwd --addr 0.0.0.0 --port 4777 \
    --sock /tmp/flexnbd.sock
  $ flexnbd status --sock /tmp/flexnbd.sock
  pid=9635 is_mirroring=false has_control=true
  
  $

Note that the status output is newline-terminated.

  MIGRATING

To migrate, we need to provide a destination file of the right size.

  $ dd if=/dev/urandom of=/tmp/data bs=1024 count=1K
  $ truncate -s 1M /tmp/data.copy
  $ flexnbd serve --file /tmp/data --addr 0.0.0.0 --port 4778 \
      --sock /tmp/flex-source.sock &
  $ flexnbd listen --file /tmp/data.copy --addr 0.0.0.0 --port 4779 \
      --sock /tmp/flex-dest.sock &
  $

Now we check the status of each server, to check that they are both in
the right state:

  $ flexnbd status --sock /tmp/flex-source.sock
  pid=9648 is_mirroring=false has_control=true
  $ flexnbd status --sock /tmp/flex-dest.sock
  pid=9651 is_mirroring=false has_control=false
  $

With this knowledge in hand, we can start the migration:

  $ flexnbd mirror --addr 127.0.0.1 --port 4779 \
      --sock /tmp/flex-source.sock
  Migration started
  [1]  + 9648 done       flexnbd serve --addr 0.0.0.0 --port 4778
  [2]  + 9651 done       flexnbd listen --addr 0.0.0.0 --port 4779
  $

Note that because the file is so small in this case, we see the source
server quit soon after we start the migration, and the destination
exited at roughly the same time.

BUGS

Should be reported on GitHub at

 * https://github.com/BytemarkHosting/flexnbd-c/issues

AUTHOR

Originally written by Alex Young <alex@blackkettle.org>.
Original concept and core code by Matthew Bloch <matthew@bytemark.co.uk>.
Proxy mode written by Nick Thomas <me@ur.gs>.

The full commit history is available on GitHub.

SEE ALSO

flexnbd-proxy(1), nbd-client(8), xnbd-server(8), xnbd-client(8)

COPYRIGHT

Copyright (c) 2012-2016 Bytemark Hosting Ltd. Free use of this
software is granted under the terms of the GNU General Public License
version 3 or later.


debian/changelog
flexnbd (0.5.0) stable; urgency=medium

  [ Patrick J Cherry ]
  * Explicitly close the server control socket, and wait for it to close, to
    prevent deadlocks during the server clean-up process (#40 !58)
  * Ensure mirroring can be restarted after a break command is sent to the
    source (#37, !59)

 -- James Carter <james.carter@bytemark.co.uk>  Fri, 11 Jan 2019 10:37:23 +0000

flexnbd (0.4.0) stable; urgency=medium

  * Ensure proxy state is completely reset before upstream init is read,
    ensuring any waiting requests are fully replayed (#39, !54)

 -- Patrick J Cherry <patrick@bytemark.co.uk>  Thu, 15 Nov 2018 14:24:26 +0000

flexnbd (0.3.0) stable; urgency=medium

  * Force a msync after every write, ignoring FUA flag, or lack thereof (!51).

 -- Patrick J Cherry <patrick@bytemark.co.uk>  Tue, 24 Apr 2018 12:05:43 +0100

flexnbd (0.2.0) stable; urgency=medium

  [ James Carter ]
  * Set TCP keepalive on sockets so broken connections are reaped (#33, !33,
    !36)
  * Add a context to logs to make debugging problems easier (#34, !34)

  [ Chris Cottam ]
  * Increased NBD_MAX_SIZE from 1MB to 32MB for qemu 2.11 (!35)

  [ Patrick J Cherry ]
  * Added FLUSH and FUA support (!38) 
  * Server returns ENOSPC in response to writes beyond the end of the
    filesystem, and EINVAL to unknown commands. (#36, !40)
  * Proxy passes all NBD protocol errors through to the client instead of
    disconnecting and retrying (#36, !40)
  * Fix struct types in readwrite.c (#35, !41)
  * Ignore ends of discs that stray outside of 512-byte sector sizes (!42). 
  * Tweak logging for readloop failures (!44)
  * Alter semantics of NBD_MAX_BLOCK_SIZE to remove struct size overheads when
    calculating if a request exceeds the max block size (!45)
  * Added tests for setting TCP_NODELAY on upstream-reconnections in the
    proxy, and refactored the other LD_PRELOAD tests (!43)
  * Clean up dead threads before calculating the number of connected clients
    on the status command (!46)

 -- Patrick J Cherry <patrick@bytemark.co.uk>  Tue, 20 Feb 2018 11:43:22 +0000

flexnbd (0.1.7) stable; urgency=medium

  * Return bytes_left in migration statistics.

 -- Chris Elsworth <chris.elsworth@bytemark.co.uk>  Fri, 14 Jul 2017 17:00:38 +0100

flexnbd (0.1.6) stable; urgency=medium

  * Remove lots of per-cpu compiler flags, notably march=native.

 -- Patrick J Cherry <patrick@bytemark.co.uk>  Thu, 13 Apr 2017 12:48:23 +0100

flexnbd (0.1.5) stable; urgency=medium

  * Switched to native packaging.

 -- Patrick J Cherry <patrick@bytemark.co.uk>  Mon, 23 Jan 2017 13:51:58 +0000

flexnbd (0.1.5-1) stable; urgency=medium

  * This fixes the compiler warning pointer-to-int-cast in serve.c.
  * Fix up "wrong" handle type from char* to uint64_t in check_readwrite test.
  * fix check_bitset test on 32-bit platforms.
  * Skip large file test on 32-bit platforms.

 -- Patrick J Cherry <patrick@bytemark.co.uk>  Fri, 07 Oct 2016 12:25:36 +0100

flexnbd (0.1.4-1) stable; urgency=medium

  [ Michel Pollet ]
  * Simplified Makefile
  * Makefile: don't include *.d's before 'all'

  [ Patrick J Cherry ]
  * Added net-tools to the build-deps for testing
  * Fixed "double-definition of constants" warning
  * Use a BINARY encoded string when doing read/write comparisons.
  * Moved all tasks from Rake to Make
  * Added dpkg-dev to build-deps
  * Updated manpages, replace a2x with txt2man
  * Fixed up internal test names (copy/pasta?)
  * Fixed check_readwrite test to pass correct handle to fd_write_reply
  * Fix up nbdtypes test to correctly use htobe64
  * Update Makefile to specify dependencies properly for tests

  [ James F. Carter ]
  * update tests to reflect changes in handle storage
  * run tests in gitlab-ci
  * force binary encoding in a ruby2.1-compatible way
  * add missing deps on server object files when building check binaries
  * remove unnecessary step in gitlab-ci
  * failures in make check now result in an error

 -- Patrick J Cherry <patrick@bytemark.co.uk>  Thu, 06 Oct 2016 15:49:33 +0100

flexnbd (0.1.3-1) stable; urgency=medium

  [ Michel Pollet ]
  * Close socket fix, might relate to migration crashing
  * Attempt at fixing bind() bug
  * Simplified NBD handle comparisons
  
  [ Patrick J Cherry ]
  * Non-maintainer upload.

 -- Patrick J Cherry <patrick@bytemark.co.uk>  Tue, 04 Oct 2016 21:20:27 +0100

flexnbd (0.1.2-1) stable; urgency=medium

  * Non-maintainer upload.
  * Updated packaging.

 -- Patrick J Cherry <patrick@bytemark.co.uk>  Tue, 30 Aug 2016 21:14:05 +0100

flexnbd (0.1.1-455) stable; urgency=low

  * flexnbd: We only require 1/8th of the memory we allocate for bitsets (bits vs. bytes confusion) [r455]

 -- nick <nick@bytemark.co.uk>  Wed, 13 May 2015 09:25:09 +0100

flexnbd (0.1.1-454) stable; urgency=low

  * Makefile: Add -lm to LLDFLAGS [r454]

 -- nick <nick@bytemark.co.uk>  Wed, 01 Apr 2015 12:39:07 +0100

flexnbd (0.1.1-453) stable; urgency=low

  * Moved the server-specific files into src/server [r453]

 -- Alex Young <alex@bytemark.co.uk>  Tue, 11 Mar 2014 11:05:43 +0000

flexnbd (0.1.1-452) stable; urgency=low

  * Remove the proxy's dependency on flexnbd.h [r452]

 -- Alex Young <alex@bytemark.co.uk>  Tue, 11 Mar 2014 10:37:00 +0000

flexnbd (0.1.1-451) stable; urgency=low

  * Tighten up some variable scopes. [r451]

 -- Alex Young <alex@bytemark.co.uk>  Tue, 11 Mar 2014 10:24:29 +0000

flexnbd (0.1.1-450) stable; urgency=low

  * Avoid a potential null pointer dereference [r450]

 -- Alex Young <alex@bytemark.co.uk>  Tue, 11 Mar 2014 09:57:19 +0000

flexnbd (0.1.1-449) stable; urgency=low

  * Fix a harmless buffer overflow [r449]

 -- Alex Young <alex@bytemark.co.uk>  Tue, 11 Mar 2014 09:49:25 +0000

flexnbd (0.1.1-448) stable; urgency=low

  * Merge [r448]

 -- Alex Young <alex@bytemark.co.uk>  Thu, 27 Feb 2014 16:18:17 +0000

flexnbd (0.1.1-447) stable; urgency=low

  * Replace off64_t with uint64_t where it makes sense to do so. [r447]


control.c
/*  FlexNBD server (C) Bytemark Hosting 2012

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/** The control server responds on a UNIX socket and services our "remote"
 *  commands which are used for changing the access control list, initiating
 *  a mirror process, or asking for status.  The protocol is pretty simple -
 *  after connecting the client sends a series of LF-terminated lines, followed
 *  by a blank line (i.e. double LF).  The first line is taken to be the command
 *  name to invoke, and the lines before the double LF are its arguments.
 *
 *  These commands can be invoked remotely from the command line, with the 
 *  client code to be found in remote.c
 */

#include "params.h"
#include "util.h"
#include "ioutil.h"
#include "parse.h"
#include "readwrite.h"
#include "bitset.h"

#include <stdlib.h>
#include <string.h>
#include <sys/un.h>
#include <unistd.h>

/** The mirror code will split NBD writes, making them this long as a maximum */
static const int mirror_longest_write = 8<<20;

/** If, during a mirror pass, we have sent this number of bytes or fewer, we
 *  go to freeze the I/O and finish it off.  This is just a guess.
 */
static const int mirror_last_pass_after_bytes_written = 100<<20;

/** The largest number of full passes we'll do - the last one will always 
 *  cause the I/O to freeze, however many bytes are left to copy.
 */
static const int mirror_maximum_passes = 7;

/** Thread launched to drive mirror process */
void* mirror_runner(void* serve_params_uncast)
{
	const int last_pass = mirror_maximum_passes-1;
	int pass;
	struct mode_serve_params *serve = (struct mode_serve_params*) serve_params_uncast;
	struct bitset_mapping *map = serve->mirror->dirty_map;
	
	for (pass=0; pass < mirror_maximum_passes; pass++) {
		uint64_t current = 0;
		uint64_t written = 0;
		
		debug("mirror start pass=%d", pass);
		
		if (pass == last_pass) {
			/* last pass, stop everything else */
			SERVER_ERROR_ON_FAILURE(
				pthread_mutex_lock(&serve->l_accept),
				"Problem with accept lock"
			);
			SERVER_ERROR_ON_FAILURE(
				pthread_mutex_lock(&serve->l_io),
				"Problem with I/O lock"
			);
		}
		
		while (current < serve->size) {
			int run;
		
			run = bitset_run_count(map, current, mirror_longest_write);
			
			debug("mirror current=%ld, run=%d", current, run);
			
			/* FIXME: we could avoid sending sparse areas of the 
			 * disc here, and probably save a lot of bandwidth and
			 * time (if we know the destination starts off zeroed).
			 */ 
			if (bitset_is_set_at(map, current)) {
				/* We've found a dirty area, send it */
				debug("^^^ writing");
				
				/* We need to stop the main thread from working 
				 * because it might corrupt the dirty map.  This
				 * is likely to slow things down but will be
				 * safe.
				 */
				if (pass < last_pass)
					SERVER_ERROR_ON_FAILURE(
						pthread_mutex_lock(&serve->l_io),
						"Problem with I/O lock"
					);
				
				/** FIXME: do something useful with bytes/second */
				
				/** FIXME: error handling code here won't unlock */
				socket_nbd_write(
					serve->mirror->client, 
					current,
					run,
					0,
					serve->mirror->mapped + current
				);
				
				/* now mark it clean */
				bitset_clear_range(map, current, run);
				
				if (pass < last_pass)
					SERVER_ERROR_ON_FAILURE(
						pthread_mutex_unlock(&serve->l_io),
						"Problem with I/O unlock"
					);
				
				written += run;
			}
			current += run;
		}
		
		/* if we've not written anything */
		if (written < mirror_last_pass_after_bytes_written)
			pass = last_pass;
	}
	
	switch (serve->mirror->action_at_finish)
	{
	case ACTION_PROXY:
		debug("proxy!");
		serve->proxy_fd = serve->mirror->client;
		/* don't close our file descriptor, we still need it! */
		break;
	case ACTION_EXIT:
		debug("exit!");
		write(serve->close_signal[1], serve, 1); /* any byte will do */
		/* fall through */
	case ACTION_NOTHING:
		debug("nothing!");
		close(serve->mirror->client);
	}
	
	free(serve->mirror->dirty_map);
	free(serve->mirror);
	serve->mirror = NULL; /* and we're gone */

	SERVER_ERROR_ON_FAILURE(
		pthread_mutex_unlock(&serve->l_accept),
		"Problem with accept unlock"
	);
	SERVER_ERROR_ON_FAILURE(
		pthread_mutex_unlock(&serve->l_io),
		"Problem with I/O unlock"
	);
	
	return NULL;
}

#define write_socket(msg) write(client->socket, (msg "\n"), strlen((msg))+1)

/** Command parser to start mirror process from socket input */
int control_mirror(struct control_params* client, int linesc, char** lines)
{
	off64_t size, remote_size;
	int fd, map_fd;
	struct mirror_status *mirror;
	union mysockaddr connect_to;
	uint64_t max_bytes_per_second;
	int action_at_finish;
	
	if (linesc < 2) {
		write_socket("1: mirror takes at least two parameters");
		return -1;
	}
	
	if (parse_ip_to_sockaddr(&connect_to.generic, lines[0]) == 0) {
		write_socket("1: bad IP address");
		return -1;
	}
	
	connect_to.v4.sin_port = atoi(lines[1]);
	if (connect_to.v4.sin_port < 0 || connect_to.v4.sin_port > 65535) {
		write_socket("1: bad IP port number");
		return -1;
	}
	connect_to.v4.sin_port = htobe16(connect_to.v4.sin_port);
	
	max_bytes_per_second = 0;
	if (linesc > 2) {
		max_bytes_per_second = atoi(lines[2]);
	}
	
	action_at_finish = ACTION_PROXY;
	if (linesc > 3) {
		if (strcmp("proxy", lines[3]) == 0)
			action_at_finish = ACTION_PROXY;
		else if (strcmp("exit", lines[3]) == 0)
			action_at_finish = ACTION_EXIT;
		else if (strcmp("nothing", lines[3]) == 0)
			action_at_finish = ACTION_NOTHING;
		else {
			write_socket("1: action must be one of 'proxy', 'exit' or 'nothing'");
			return -1;
		}
	}
	
	if (linesc > 4) {
		write_socket("1: unrecognised parameters to mirror");
		return -1;
	}
	
	fd = socket_connect(&connect_to.generic);
	
	remote_size = socket_nbd_read_hello(fd);
	remote_size = remote_size; // shush compiler
	
	mirror = xmalloc(sizeof(struct mirror_status));
	mirror->client = fd;
	mirror->max_bytes_per_second = max_bytes_per_second;
	mirror->action_at_finish = action_at_finish;
	
	CLIENT_ERROR_ON_FAILURE(
		open_and_mmap(
			client->serve->filename, 
			&map_fd,
			&size, 
			(void**) &mirror->mapped
		),
		"Failed to open and mmap %s",
		client->serve->filename
	);
	
	mirror->dirty_map = bitset_alloc(size, 4096);
	bitset_set_range(mirror->dirty_map, 0, size);
	
	client->serve->mirror = mirror;
	
	CLIENT_ERROR_ON_FAILURE( /* FIXME should free mirror on error */
		pthread_create(
			&mirror->thread, 
			NULL, 
			mirror_runner, 
			client->serve
		),
		"Failed to create mirror thread"
	);
	
	write_socket("0: mirror started");
	
	return 0;
}

/** Command parser to alter access control list from socket input */
int control_acl(struct control_params* client, int linesc, char** lines)
{
	int parsed;
	struct ip_and_mask (*acl)[], (*old_acl)[];
	
	parsed = parse_acl(&acl, linesc, lines);
	if (parsed != linesc) {
		write(client->socket, "1: bad spec: ", 13);
		write(client->socket, lines[parsed], 
		  strlen(lines[parsed]));
		write(client->socket, "\n", 1);
		free(acl);
	}
	else {
		old_acl = client->serve->acl;
		client->serve->acl = acl;
		client->serve->acl_entries = linesc;
		free(old_acl);
		write_socket("0: updated");
	}
	
	return 0;
}

/** FIXME: add some useful statistics */
int control_status(struct control_params* client, int linesc, char** lines)
{
	return 0;
}

/** Master command parser for control socket connections, delegates quickly */
void* control_serve(void* client_uncast)
{
	struct control_params* client = (struct control_params*) client_uncast;
	char **lines = NULL;
	int finished=0;
	
	while (!finished) {
		int i, linesc;		
		linesc = read_lines_until_blankline(client->socket, 256, &lines);
		
		if (linesc < 1)
		{
			write(client->socket, "9: missing command\n", 19);
			finished = 1;
			/* ignore failure */
		}
		else if (strcmp(lines[0], "acl") == 0) {
			if (control_acl(client, linesc-1, lines+1) < 0)
				finished = 1;
		}
		else if (strcmp(lines[0], "mirror") == 0) {
			if (control_mirror(client, linesc-1, lines+1) < 0)
				finished = 1;
		}
		else if (strcmp(lines[0], "status") == 0) {
			if (control_status(client, linesc-1, lines+1) < 0)
				finished = 1;
		}
		else {
			write(client->socket, "10: unknown command\n", 23);
			finished = 1;
		}
		
		for (i=0; i<linesc; i++)
			free(lines[i]);
		free(lines);
	}
	
	close(client->socket);
	free(client);
	return NULL;
}

void accept_control_connection(struct mode_serve_params* params, int client_fd, union mysockaddr* client_address)
{
	pthread_t control_thread;
	struct control_params* control_params;
	
	control_params = xmalloc(sizeof(struct control_params));
	control_params->socket = client_fd;
	control_params->serve = params;

	SERVER_ERROR_ON_FAILURE(
		pthread_create(
			&control_thread, 
			NULL, 
			control_serve, 
			control_params
		),
		"Failed to create client thread"
	);
}

void serve_open_control_socket(struct mode_serve_params* params)
{
	struct sockaddr_un bind_address;
	
	if (!params->control_socket_name)
		return;

	params->control = socket(AF_UNIX, SOCK_STREAM, 0);
	SERVER_ERROR_ON_FAILURE(params->control,
	  "Couldn't create control socket");
	
	memset(&bind_address, 0, sizeof(bind_address));
	bind_address.sun_family = AF_UNIX;
	strncpy(bind_address.sun_path, params->control_socket_name, sizeof(bind_address.sun_path)-1);
	
	unlink(params->control_socket_name); /* ignore failure */
	
	SERVER_ERROR_ON_FAILURE(
		bind(params->control, &bind_address, sizeof(bind_address)),
		"Couldn't bind control socket to %s",
		params->control_socket_name
	);
	
	SERVER_ERROR_ON_FAILURE(
		listen(params->control, 5),
		"Couldn't listen on control socket"
	);
}

